credentials/sts/sts.go

```
@@ -0,0 +1,397 @@
+func (c *callCreds) metadataFromCachedToken() map[string]string {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.cachedToken == nil {
+		return nil
+	}
+
+	now := time.Now()
+	// If the cached token has not expired and the lifetime remaining on that
+	// token is greater than the minimum value we are willing to accept, go
+	// ahead and use it.
+	if c.cachedToken.expiryTime.After(now) && c.cachedToken.expiryTime.Sub(now) > minCachedTokenLifetime {
+		return map[string]string{
+			"Authorization": fmt.Sprintf("%s %s", c.cachedToken.tokenType, c.cachedToken.token),
+		}
+	}
+	return nil
+}
+
+// constructRequest creates the STS request body in JSON based on the provided
+// options.
+// - Contents of the subjectToken are read from the file specified in
+//   options. If we encounter an error here, we bail out.
+// - Contents of the actorToken are read from the file specified in options.
+//   If we encounter an error here, we ignore this field because this is
+//   optional.
+// - Most of the other fields in the request come directly from options.
+//
+// A new HTTP request is created by calling http.NewRequestWithContext() and
+// passing the provided context, thereby enforcing any timeouts specified in
+// the latter.
+func constructRequest(ctx context.Context, opts Options) (*http.Request, error) {
+	subToken, err := readSubjectTokenFrom(opts.SubjectTokenPath)
+	if err != nil {
+		return nil, err
+	}
+	reqScope := opts.Scope
+	if reqScope == "" {
+		reqScope = defaultCloudPlatformScope
+	}
+	reqParams := &RequestParameters{
+		GrantType:          tokenExchangeGrantType,
+		Resource:           opts.Resource,
+		Audience:           opts.Audience,
+		Scope:              reqScope,
+		RequestedTokenType: opts.RequestedTokenType,
+		SubjectToken:       string(subToken),
+		SubjectTokenType:   opts.SubjectTokenType,
+	}
+	if opts.ActorTokenPath != "" {
+		actorToken, err := readActorTokenFrom(opts.ActorTokenPath)
+		if err == nil {
```

Do you intend to throw away non-nil errors?

---

Earlier I wasn't checking for `opts.ActorTokenPath != ""`. So, I was ignoring errors. Now that I'm checking for it, I shouldn't be ignoring errors. Fixed. Thanks.